torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1441,8 +1442,184 @@ void InitXlaModuleBindings(py::module m) {
   });
 
   BuildProfilerSubmodule(&m);
-}
 
+  m.def("_get_tensors_handle",
+        [](const std::vector<at::Tensor>& tensors) -> std::vector<int64_t> {
+          std::vector<int64_t> handles;
+          handles.reserve(tensors.size());
+          for (auto& tensor : tensors) {
+            handles.push_back(bridge::GetXlaTensor(tensor)->GetOpaqueHandle());
+          }
+          return handles;
+        });
+
+  // -------------Dynamo Integration API Start-------------------------
+  /*
+   * Return tensor ids and tensors for DeviceData nodes.
+   */
+  m.def("_get_tensors_xla_device_data_node",
+        [](const std::vector<at::Tensor>& tensors)
+            -> std::pair<std::vector<int64_t>, std::vector<at::IValue>> {
+          std::vector<int64_t> tensor_ids;
+          std::vector<at::IValue> ivalues;
+          std::vector<torch::lazy::Node*> roots;
+          for (auto& tensor : tensors) {
+            auto xtensor = bridge::TryGetXlaTensor(tensor);
+            roots.push_back(xtensor->GetIrValue().node.get());
+          }
+          auto post_order = Util::ComputePostOrder(roots);
+          std::unordered_set<xla::ComputationClient::Data::OpaqueHandle>
+              data_handles;
+
+          for (auto nodeptr : post_order) {
+            const torch_xla::DeviceData* device_data =
+                torch_xla::DeviceData::Cast(nodeptr);
+            if (!device_data) {
+              continue;
+            }
+
+            const auto backend_data = device_data->data();
+
+            // follow the implementation of XLATensor::CreateTensorNode
+            auto dataptr =
+                ((torch_xla::XLAData*)backend_data.get())->xla_data();
+            TORCH_CHECK(dataptr);
+            auto infoptr = (torch_xla::DeviceDataInfo*)dataptr->info();
+            TORCH_CHECK(infoptr);
+
+            // Dedup by handle
+            auto handle = backend_data->GetHandle();
+            if (!data_handles.insert(handle).second) {
+              continue;
+            }
+
+            // add tensor_id after we make sure the handle does not exist yet.
+            tensor_ids.push_back(infoptr->tensor_id);
+            // This will create a new XLATensor with different Id which is not
+            // ideal. However there is not an easy way to get the origional
+            // tensor with the tensor_id.
+            // TODO(JackCaoG): invesgate if we have idle XLATensor.
+            auto tensor = bridge::AtenFromXlaTensor(
```

Comment:
TODO, I think we can just create a map at the beginning of this method and reuse the pass in `at::Tensor` instead of creating new ones.

Reply:
OK I am wrong. The tensor here might not be part of the input tensors. For example
```
    t1 = torch.randn(20, 5).to(xla_device)
    t2 = torch.randn(20, 5).to(xla_device)
    t3 = t2 + t1
    t4 = t3 * t2
    res_pair = torch_xla._XLAC._get_tensors_xla_device_data_node([t4])
```

both `t1` and `t2` are device_datas but they are not part of input passed to `_get_tensors_xla_device_data_node`

Reply:
actually this is OK, when I look at `RegisterTensor`, when it sees same `XlaData` it will keep the TensorID